File: java/kvdkjni/configs.cc
/* SPDX-License-Identifier: BSD-3-Clause
 * Copyright(c) 2021-2022 Intel Corporation
 */

#include <assert.h>

#include "include/io_pmem_kvdk_Configs.h"
#include "kvdkjni/kvdkjni.h"


/**
 * Class:     io_pmem_kvdk_Configs
 * Method:    newConfigs
 * Signature: ()J
 */
jlong Java_io_pmem_kvdk_Configs_newConfigs(
    JNIEnv*, jclass) {
  auto* cfg = new kvdk::Configs();
  return GET_CPLUSPLUS_POINTER(cfg);
}

/**
 * Class:     io_pmem_kvdk_Configs
 * Method:    setPMemFileSize
 * Signature: (JJ)V
 */
void Java_io_pmem_kvdk_Configs_setPMemFileSize(
  JNIEnv*, jobject, jlong handle, jlong size) {
  reinterpret_cast<kvdk::Configs *>(handle)->pmem_file_size = size;
}

/**
 * Class:     io_pmem_kvdk_Configs
 * Method:    setPMemSegmentBlocks
 * Signature: (JJ)V
 */
void Java_io_pmem_kvdk_Configs_setPMemSegmentBlocks(
  JNIEnv*, jobject, jlong handle, jlong blocks) {
  reinterpret_cast<kvdk::Configs *>(handle)->pmem_segment_blocks = blocks;
}

/**
 * Class:     io_pmem_kvdk_Configs
 * Method:    setHashBucketNum
 * Signature: (JJ)V
 */
void Java_io_pmem_kvdk_Configs_setHashBucketNum(
  JNIEnv*, jobject, jlong handle, jlong num) {
  reinterpret_cast<kvdk::Configs *>(handle)->hash_bucket_num = num;
}

/**
 * Class:     io_pmem_kvdk_Configs
 * Method:    closeInternal
 * Signature: (J)V
 */
void Java_io_pmem_kvdk_Configs_closeInternal(
    JNIEnv*, jobject, jlong handle) {
  auto* cfg = reinterpret_cast<kvdk::Configs *>(handle);
  assert(cfg != nullptr);
  delete cfg;
}

File: volatile/java/src/main/java/io/pmem/kvdk/Configs.java
/* SPDX-License-Identifier: BSD-3-Clause
 * Copyright(c) 2021-2022 Intel Corporation
 */

package io.pmem.kvdk;

/** KVDK engine configs. */
public class Configs extends KVDKObject {
    static {
        Engine.loadLibrary();
    }

    public Configs() {
        super(newConfigs());
    }

    public Configs setMaxAccessThreads(final long num_threads) {
        setMaxAccessThreads(nativeHandle_, num_threads);
        return this;
    }

    public Configs setHashBucketNum(final long num) {
        setHashBucketNum(nativeHandle_, num);
        return this;
    }

    public Configs setDestMemoryNodes(final String nodes) {
        setDestMemoryNodes(nativeHandle_, nodes);
        return this;
    }

    public Configs setOptLargeSortedCollectionRecovery(final boolean opt) {
        setOptLargeSortedCollectionRecovery(nativeHandle_, opt);
        return this;
    }

    public Configs setUseDevDaxMode(final boolean use) {
        setUseDevDaxMode(nativeHandle_, use);
        return this;
    }

    public Configs setCleanThreads(final long num_threads) {
        setCleanThreads(nativeHandle_, num_threads);
        return this;
    }

    // Native methods
    private static native long newConfigs();

    private native void setMaxAccessThreads(long handle, long num);

    private native void setHashBucketNum(long handle, long num);

    private native void setDestMemoryNodes(long handle, String nodes);

    private native void setOptLargeSortedCollectionRecovery(long handle, boolean opt);

    private native void setUseDevDaxMode(long handle, boolean use);

    private native void setCleanThreads(long handle, long num);

    @Override
    protected final native void closeInternal(long handle);
}
